akka-http/src/main/scala/akka/http/scaladsl/server/directives/ReverseProxyDirectives.scala

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright (C) 2009-2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+
+package akka.http.scaladsl.server.directives
+
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.Http.{ HostConnectionPool, OutgoingConnection }
+import akka.http.scaladsl.model.Uri.Authority
+import akka.http.scaladsl.model.{ HttpResponse, _ }
+import akka.http.scaladsl.model.headers._
+import akka.http.scaladsl.server._
+import akka.http.scaladsl.server.directives.BasicDirectives._
+import akka.http.scaladsl.server.directives.FutureDirectives._
+import akka.http.scaladsl.server.directives.RouteDirectives._
+import akka.pattern.CircuitBreaker
+import akka.stream._
+import akka.stream.scaladsl.{ Flow, Keep, Sink, Source }
+
+import scala.collection.immutable.Seq
+import scala.concurrent.Future
+import scala.util.{ Failure, Success, Try }
+
+trait ReverseProxyDirectives {
+  type RequestInterceptor = (HttpRequest) ⇒ HttpRequest
+  type ResponseInterceptor = ((HttpResponse) ⇒ HttpResponse)
+  type RequestExecutor = ((HttpRequest) ⇒ Future[HttpResponse])
+
+  def completeUnmatchedPathWithReverseProxy(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None, responseInterceptor: ResponseInterceptor = identity[HttpResponse](_))(implicit system: ActorSystem): Route = {
+    forward(authority, circuitBreakerOption).tapply(x ⇒ complete(x._1.map(responseInterceptor)))
+  }
+
+  def forward(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
+    extractUri.flatMap { uri ⇒
+      val forwardUri: Uri = uri.withAuthority(authority)
+      forwardToUri(forwardUri, circuitBreakerOption)
+    }
+  }
+
+  def forwardUnmatchedPath(authority: Authority, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
+    extractUnmatchedPath.flatMap { unmatched ⇒
+      extractUri.flatMap { uri ⇒
+        val forwardUri: Uri = uri.withAuthority(authority).withPath(unmatched)
+        forwardToUri(forwardUri, circuitBreakerOption)
+      }
+    }
+  }
+
+  def forwardToUri(uri: Uri, circuitBreakerOption: Option[CircuitBreaker] = None)(implicit system: ActorSystem): Directive1[Try[HttpResponse]] = {
```

Comment:
Well, to make a request to another (remote) server, we either need an `ActorSystem` (implicit or explicit) to construct a `Function` `(HttpRequest) => Future[HttpResponse]` or a `Function` `(HttpRequest) => Future[HttpResponse]` directly.

If I get you right, you'd prefer to explicitly pass a `ReverseProxyConfig` to the Directives:
```
trait ReverseProxyConfig {
  val circuitBreakerOption: Option[CircuitBreaker]
  val requestExecutor: (HttpRequest) => Future[HttpResponse]
}
```

But how can we provide a default `requestExecutor`, if we don't have an `ActorSystem` in scope?
If the `requestExecutor` should be configurable via Typesafe Config, how can we construct a Function `(HttpRequest) => Future[HttpResponse]` without an `ActorSystem` in implicit scope?